### ISSUE TITLE
Command 'CMake: Quick Start' not honoring workspace cmake.sourceDirectory

### ISSUE THREAD

**Issue by keithel** [microsoft/vscode-cmake-tools#2145]
### Brief Issue Summary

When I try to create a new `CMakeLists.txt` file for a new project, I find that it fails because it is trying to access files in `${workspaceFolder}/Maya` - which happens to be the `cmake.sourceDirectory` that is set at the User level.
This particular project was created new, so the `cmake.sourceDirectory` for the workspace was set to just `${workspaceFolder}`.

It appears that the `CMake: Quick Start` feature is not honoring the workspace `cmake.sourceDirectory` setting.

The error I am seeing is:
`[rollbar] Unhandled exception: Unhandled Promise rejection: quickStart Error: ENOENT: no such file or directory, open '/home/myusername/Build/Autodesk/maya/test-qopenglwindow/Maya/main.cpp' {}
[main] Configuring folder: test-qopenglwindow `

User `cmake.sourceDirectory`: `${workspaceFolder}/Maya`
Workspace `cmake.sourceDirectory`: `${workspaceFolder}`

This occurs with CMake Tools v1.8.1 running on centOS 7.9.2009, cmake 3.17.5.

Note: I have replaced my username with `myusername` in the logs below. Everything else remains the same.

### CMake Tools Diagnostics

`>CMake: Log Diagnostics` command search failed. Instead, below is the output of the `CMake/Build` log I can access from the Output pane.

```
<!-- Paste the `CMake: Log Diagnostics` output HERE -->
[rollbar] Unhandled exception: Unhandled Promise rejection: quickStart Error: ENOENT: no such file or directory, open '/home/myusername/Build/Autodesk/maya/test-qopenglwindow/Maya/main.cpp' {}
[main] Configuring folder: test-qopenglwindow 
```

### Debug Log

```
[main] Safe constructing new CMakeTools instance
[variant] Constructing VariantManager
[main] Constructing new CMakeTools instance
[main] Starting CMakeTools second-phase init
[rollbar] "sourceDirectory" is not a directory {"sourceDirectory":"/home/myusername/Build/Autodesk/maya/test-qopenglwindow/Maya"}
[variant] Loaded new set of variants
[main] CMakeTools instance initialization complete.
[kit] Reading kits file /home/myusername/.local/share/CMakeTools/cmake-tools-kits.json
[kit] Successfully loaded 2 kits from /home/myusername/.local/share/CMakeTools/cmake-tools-kits.json
[kit] Not reading non-existent kits file: /home/myusername/Build/Autodesk/maya/test-qopenglwindow/.vscode/cmake-kits.json
[rollbar] "sourceDirectory" is not a directory {"sourceDirectory":"/home/myusername/Build/Autodesk/maya/test-qopenglwindow/Maya"}
[main] Injecting new Kit into CMake driver
[extension] [6945] cmake.buildType started
[extension] [6945] cmake.buildType finished (returned "Debug")
[extension] [6692] cmake.buildDirectory started
[main] Starting new CMake driver
[main] Starting CMake driver
[main] Using generator from user configuration: Ninja
[cmakefileapi-driver] Creating instance of CMakeFileApiDriver
[driver] Initializating base driver using kit
[kit] Testing GCC binary: /opt/rh/devtoolset-9/root/usr/bin/x86_64-redhat-linux-gcc-9
[proc] Executing command: /opt/rh/devtoolset-9/root/usr/bin/x86_64-redhat-linux-gcc-9 -v
[kit] Detected GCC compiler: /opt/rh/devtoolset-9/root/usr/bin/x86_64-redhat-linux-gcc-9
[driver] CMakeDriver Kit set to GCC 9.3.1 x86_64-redhat-linux
[driver] Run _refreshExpansions
[driver] Run _refreshExpansions cb
[rollbar] "sourceDirectory" is not a directory {"sourceDirectory":"/home/myusername/Build/Autodesk/maya/test-qopenglwindow/Maya"}
[driver] Setting new variant Debug
[driver] Run _refreshExpansions
[driver] Run _refreshExpansions cb
[rollbar] "sourceDirectory" is not a directory {"sourceDirectory":"/home/myusername/Build/Autodesk/maya/test-qopenglwindow/Maya"}
[extension] [6692] cmake.buildDirectory finished (returned "/home/myusername/Build/Autodesk/builds/test-qopenglwindow/maya/Debug")
[cache] Reading CMake cache file /home/myusername/Build/Autodesk/builds/test-qopenglwindow/maya/Debug/CMakeCache.txt
[cache] Cache file does not exist: Returning empty cache data
[extension] [8213] cmake.quickStart started
[rollbar] Unhandled exception: Unhandled Promise rejection: quickStart Error: ENOENT: no such file or directory, open '/home/myusername/Build/Autodesk/maya/test-qopenglwindow/Maya/main.cpp' {}
[rollbar] Unhandled exception: Unhandled Promise rejection: Post-folder-open Error: Running the contributed command: 'cmake.quickStart' failed. {"folder":{"uri":{"$mid":1,"fsPath":"/home/myusername/Build/Autodesk/maya/test-qopenglwindow","external":"file:///home/myusername/Build/Autodesk/maya/test-qopenglwindow","path":"/home/myusername/Build/Autodesk/maya/test-qopenglwindow","scheme":"file"},"name":"test-qopenglwindow","index":0}}
```

**Comment by andreeis** [comment]
@keithel, thank you for creating this report. I just played with the command while having "cmake.sourceDirectory" set at the user level and there are plenty of scenarios to cover (single/multi root, run command from empty or existing folder, run from palette or via the "Create/Locate/Ignore" bottom right popup, ... etc ...). I reproduce various issues.

Just to make sure we don't overlook your exact scenario, describe please the structure of your project.
Also, does it still happen after you reload the project that was just created with the quick start command? I understand why the error happens at the beginning but it shouldn't after you define "cmake.sourceDirectory" at the project level and you reload.

Having a setting like "cmake.sourceDirectory" defined at the user level is not a scenario we thought about. Why do you need that?

**Comment by keithel** [comment]
This was with a brand new, simple project with just a single source file (I was testing something). Ultimately it will include Qt, but for these tests, it didn't - think of it as just a simple Hello World application that only depends on the standard C++ runtime libs.

I was able to make it work after resetting the User level cmake.sourceDirectory setting to ${workspaceFolder}. After I did that, I could create the project with the `CMake: Quick Start` command. Once project has been created, I could reset cmake.sourceDirectory in User settings back to what it was, and this new project still will load and build properly with CMake tools.

This occurs no matter if I just have VSCode open a new folder to the root of this project without existing CMakeLists.txt project and try creating a basic CMakeLists.txt file with CMake: Quick Start or if I actually save a new VSCode workspace file in this directory before invoking CMake: Quick Start.

Again, if I go and make the User settings cmake.sourceDirectory point to ${workspaceFolder}, then creating a CMakeLists.txt file works fine. Once created, I can put it back to what it was previously.
